### DISCUSSION TITLE
Multicopter MotorModel Plugin and PID Controller for UAV in ROS2

### DISCUSSION THREAD

**Discussion by omr-web** [bitcraze/discussions#568] [Q&A]
hello everyone,
I am making an altitude and attitude controller for an unmanned aerial vehicle (UAV, drone) with Ignition Gazebo and ROS2. I use the multicopter motor model plugin to do this. I would like your help with an interesting situation I encountered:
When I give the command to go up to my model (drone) with the joystick, the model goes up and hover in the air as I expected. But all of a sudden the model falls to the ground. The model does not move upwards even if I send the ascend command. I think this situation is not caused by the controller I wrote. I would like to explain the reason for this graphically:

https://drive.google.com/file/d/178Reu0bQNrdhItp2T65bAHIVXn2gucPa/view?usp=sharing

Above are two graphs with the same x-axis. In the graphic above, there are the target distance (setpoint), the location (measured) and the difference between the setpoint and the measured (error) for the model to go, and the velocity values ​​to go to the motors produced by the altitude controller in the graphic at the bottom. When the red graph is examined, while the controller is working correctly, that is, the model with reduced error reaches the desired (setpoint) position and drops down at a time.

When the controller graphic is examined at first, it seems to work properly, because as the error decreases (blue line in the upper graph), the velocity value that will go to the produced motors decreases. so I think the source of the problem is not the controller, but the gazebo world. Is there a factor in the gazebo world that I should pay attention to? For example, type of atmosphere, air pressure etc.

I changed the pid coefficients in the attitude controller. I observed that when the coefficients are 0, that is, when the attitude controller is disabled, the model hover in the air and falls suddenly again. The reason I disabled the Attitude controller was the possibility that the pid results produced by this attitude controller could be dropping the model while it was hovering.

this is the link of the simulation video:
https://drive.google.com/file/d/1-wcMwoHd61y1bFJst3OIfsTULB25UVhP/view?usp=sharing

Thank you for your help,
Kind regards.

**Comment by knmcguire** [comment]
Hi! Some additional questions for this:

- Are you re-using the plugin in this repository? https://github.com/bitcraze/crazyflie-simulation
- Since you have written the controller yourself it is difficult for us to determine what is going on... have you by any chance looked at the pid_gazebo branch in the repository? https://github.com/bitcraze/crazyflie-simulation/tree/pid_gazebo. There the PID controller also fails but perhaps you can check what is going wrong
- Have you tried the same controller on the webots simulation?

**Reply by omr-web** [reply]
I  want to answer respectively:
1) Yes I use this plugin  [https://github.com/bitcraze/crazyflie-simulation ](url)
2) Even when I give a fixed throttle when there is no controller, it goes up to a certain height and the drone falls from this height. The maximum height it can reach varies according to the fixed throttle value given. I don't think the problem is from the controller. so I think it is due to some feature of the world.Have you encountered such a problem in an ignition gazebo?
3) I have to use gazebo so I don't look webots

**Reply by knmcguire** [reply]
Hmm oke, it dependes on the physics engine, but if you are only giving full thruttle with no attitude controller, then it's up to physics that are on. Perhaps there is some simulated wind, or perhaps the accumulated error of the flow dynamics of the propeller. That's unfortunately all I know...

You could try if you can find any examples of other flying platforms in gazebo (new not classic) ? The px4 autopilot has it listed as one as their simulators, but we have mostly focused on webots because of their python support

**Reply by omr-web** [reply]
I'm having a little difficulty because the resources about the ignition gazebo are scarce. But thank you for your words and your help.
